### PR TITLE
I56 timeout

### LINKED ISSUE #56: Add timeout option
Provide a (crude) mechanism for ending individual location processing after a timeout. 
https://www.rdocumentation.org/packages/R.utils/versions/2.9.2/topics/withTimeout

Note (re: stan code keeping running):
"More precisely, if a function is implemented in native code (e.g. C) and the developer of that function does not check for user interrupts, then you cannot interrupt that function neither via a user interrupt (e.g. Ctrl-C) nor via the built-in time out mechanism. To change this, you need to contact the developer of that piece of code and ask them to check for R user interrupts in their native code."
I have found reference to Rstan checking for interrupts every iteration so we'll have to see what happens.

### PR DESCRIPTION
**Pull request by joeHickson**
closes #56 

- [x] reset return_timings and max_execution_time defaults once testing is complete

### CODE REVIEW

**Comment by joeHickson** [comment]
I think this is ready for review. 
Improved logging is also in place (in line with epiforecasts/covid-rt-estimates#20 and epiforecasts/covid-rt-estimates#17 )

To test - set max_execution_time to something suitable to get a mixed result. I have been using canada and 1/2 hr but that will be machine specific. 

@seabbs  please can you confirm that the summary results are acceptable with missing locations. I have pushed them to a temporary branch for verification:
https://github.com/epiforecasts/covid-rt-estimates/tree/i56_timeout_test_results/subnational/canada/cases/summary

Sample of the log below:

WARN [2020-09-02 11:26:02] Quebec: Examine the pairs() plot to diagnose sampling problems
 - 
TRACE [2020-09-02 11:26:13] epinow returned for region Quebec
INFO [2020-09-02 11:26:13] Completed estimates for: Quebec
WARN [2020-09-02 11:26:34] Ontario: There were 18 divergent transitions after warmup. See
http://mc-stan.org/misc/warnings.html#divergent-transitions-after-warmup
to find out why this is a problem and how to eliminate them. - 
WARN [2020-09-02 11:26:34] Ontario: Examine the pairs() plot to diagnose sampling problems - 
WARN [2020-09-02 11:26:37] Ontario: Bulk Effective Samples Size (ESS) is too low, indicating posterior means and medians may be unreliable.
Running the chains for more iterations may help. See http://mc-stan.org/misc/warnings.html#bulk-ess -
TRACE [2020-09-02 11:26:45] epinow returned for region Ontario
INFO [2020-09-02 11:26:45] Completed estimates for: Ontario
WARN [2020-09-02 11:30:20] region Saskatchewan timed out
TRACE [2020-09-02 11:30:20] epinow returned for region Saskatchewan
INFO [2020-09-02 11:30:20] Completed estimates for: Saskatchewan
WARN [2020-09-02 11:31:13] region Manitoba timed out
TRACE [2020-09-02 11:31:13] epinow returned for region Manitoba
INFO [2020-09-02 11:31:13] Completed estimates for: Manitoba
TRACE [2020-09-02 11:31:13] processing errors
TRACE [2020-09-02 11:31:13] 2 runtime errors caught
WARN [2020-09-02 11:31:13] Location Manitoba killed due to timeout
WARN [2020-09-02 11:31:13] Location Saskatchewan killed due to timeout
TRACE [2020-09-02 11:31:13] Calling regional_summary
TRACE [2020-09-02 11:31:14] Getting regional results
TRACE [2020-09-02 11:31:14] Summarising results
DEBUG [2020-09-02 11:31:24] resetting future plan to sequential

File: R/epinow.R [hunk 1]
```diff
@@ -346,94 +349,128 @@ regional_epinow <- function(reported_cases,
...
-  
+
   ## Exclude zero regions
   reported_cases <- reported_cases[!is.na(region)][region %in% eval_regions]
-  
+
   futile.logger::flog.info("Producing estimates for: %s",
-          paste(eval_regions, collapse = ", "))
-  
+                           paste(eval_regions, collapse = ", "))
```
Comment (seabbs) [inline R/epinow.R:381]: :)

File: R/epinow.R [hunk 2]
```diff
@@ -346,94 +349,128 @@ regional_epinow <- function(reported_cases,
...
+          withCallingHandlers(EpiNow2::epinow(
+            reported_cases = regional_cases,
+            target_folder = target_folder,
+            target_date = target_date,
+            return_estimates = TRUE,
+            cores = cores,
+            ...),
+                              warning = function(w) {
+                                futile.logger::flog.warn("%s: %s - %s", target_region, w$message, toString(w$call))
+                                cnd_muffle(w)
```
Comment (seabbs) [inline R/epinow.R:417]: suggest call package to fit style (but not required)

File: R/epinow.R [hunk 3]
```diff
@@ -346,94 +349,128 @@ regional_epinow <- function(reported_cases,
...
     }
+  }
 
-  regional_out <- purrr::map(regional_out, ~ .$result)
-  names(regional_out) <- regions
-  
-  
-  if (summary) {
+  regional_out <- purrr::map(regional_out, ~.$result)
+  sucessful_regional_out <- purrr::keep(purrr::compact(regional_out), function(row) is.finite(row$timings))
```
Comment (seabbs) [inline R/epinow.R:464]: need to add to @importFrom purrr

File: R/epinow.R [hunk 4]
```diff
@@ -346,94 +349,128 @@ regional_epinow <- function(reported_cases,
...
+        ),
+        TimeoutException = function(ex) {
+          futile.logger::flog.warn("region %s timed out", target_region)
+          return(list("timings" = Inf))
+        }
+      )
+    )
+    futile.logger::flog.trace("epinow returned for region %s", target_region)
+    if (!exists("timings", out)) { # only exists if it failed and is Inf
+      out$timings = timing['elapsed']
```
Comment (seabbs) [inline R/epinow.R:430]: So this is nested in each region (so out$regional_out$france$timings) but will only ever have one entry. Perhaps it should therefore be $timing or be mapped to out$timings instead (and would then need the region as a name). Happy with either or staying as is.

**Review by seabbs — CHANGES_REQUESTED** [review]
Looks good. Just a few small notes and a redoc needed (https://github.com/epiforecasts/EpiNow2/pull/63/checks?check_run_id=1061264418). Other notes are unrelated and not in fact issues.

Missing locations are okay and a sensible fall back (though in a real run we would probably want to increase the run time if we are seeing this a lot (obviously)). 

Sample warning, trace, progress messages look good but there is a bit of duplication. I don't know enough about good logging practises to say if this is desirable (repeat warning messages, warning message followed by a claim that estimates are being returned). Happy as is.

File: R/epinow.R [hunk 5]
```diff
@@ -67,218 +67,217 @@
...
-   }
-   
-#  # Plot --------------------------------------------------------------------
-
-   plots <- report_plots(summarised_estimates = estimates$summarised,
-                         reported = reported_cases, target_folder = target_folder)
-
- # Copy all results to latest folder ---------------------------------------
-  if (!is.null(target_folder)) {  
+  # Forecast infections and reproduction number -----------------------------
```
[the remark this answers is not in the document]
Comment (joeHickson) [reply R/epinow.R:167]: might be merge noise, let me check

[the remark this answers is not in the document]
Comment (joeHickson) [reply R/epinow.R:167]: looks like it's just a big block of whitespace change - did an autoformat at some point in the process and it shuffled everything... Telling github to ignore whitespace changes hides this block as no change.

[the remark this answers is not in the document]
Comment (joeHickson) [reply R/epinow.R:167]: You have me paranoid now! let me cut a new branch and confirm the same thing happens....

[the remark this answers is not in the document]
Comment (joeHickson) [reply R/epinow.R:167]: https://github.com/epiforecasts/EpiNow2/compare/i56_timeout_2

Do you use autoformatter in an IDE to reflow your code? I might need to configure PyCharm to reformat according to the correct style.

File: R/epinow.R (hunk 2 above)
Comment (joeHickson) [reply R/epinow.R:417]: I'm not sure I understand, would you mind expanding?

**Comment by joeHickson** [comment]
@seabbs 
I have clarified the logging so that you should only get warning for timeout, info for good completion and no trace (it was a debug line from when I was having issues). This should give a cleaner output:

WARN [2020-09-02 11:26:37] Ontario: Bulk Effect...
INFO [2020-09-02 11:26:45] Completed estimates for: Ontario <- no need for a trace straight before a complete message
WARN [2020-09-02 11:30:20] region Saskatchewan timed out
WARN [2020-09-02 11:31:13] region Manitoba timed out <- if it's timed out you don't get the completion message
TRACE [2020-09-02 11:31:13] processing errors <- no longer repeats the timeout warnings

I have re-introduced the default value on the timeout for testing so don't merge until I get results (should be about 1800 seconds or so, then I'll whip it back out ;) )

File: R/epinow.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (seabbs) [reply R/epinow.R:167]: I am just using the Rstudio defaults and to me, this mostly looks like my sloppiness. Your automatic changes all look better actually so no problem.

[the remark this answers is not in the document]
Comment (seabbs) [reply R/epinow.R:167]: ah I did not know you good get it to ignore whitespace - that is magical!

File: R/epinow.R (hunk 2 above)
Comment (seabbs) [reply R/epinow.R:417]: as in everywhere else we have used `purrr::map` so here it should really be `rlang::cnd_muffle` to match. That approaches has issues though as code can pass checks even when a function is missing from the NAMESPACE (like `purrr::keep` in this PR) so TBH may not be ideal practice.

**Comment by seabbs** [comment]
Could you please also increment the package version in the DESCRIPTION to `1.2.0` and add this feature to a new news section in `NEWS.md`. `1.1.0` just hit CRAN so need to make it clear this feature is not in that version.

**Comment by seabbs** [comment]
Thanks looks really great - this will be a great help! Just waiting for CI and then will merge.

**Comment by joeHickson** [comment]
I'm just waiting on a final run of canada to complete too

**Comment by joeHickson** [comment]
That failed, I'll need to look at it tomorrow. I'll revert to draft whilst I investigate.

**Comment by seabbs** [comment]
I see region_timings -> region_timings needing to be updated in the docs.

**Comment by joeHickson** [comment]
> I see region_timings -> region_timing needing to be updated in the docs.
you caught me mid-push!

**Comment by joeHickson** [comment]
package check is failing because episoon is not available on the windows build. @seabbs  any ideas?

**Comment by seabbs** [comment]
It looks like the issue is with a failure to install the `vctrs` 📦 . I think that is a dependency of a dependency and likely due to a CRAN update from the developers....

**Comment by seabbs** [comment]
Yup looks like recently pushed to CRAN: https://cran.r-project.org/web/packages/vctrs/index.html

**Comment by joeHickson** [comment]
Annoying that windows is the only one affected.

**Review by seabbs — APPROVED** [review]
Looks great. GitHub docs action is a great idea. Thanks for adding this.

**Pull request merged**